src/modules/cdp.ts

```diff
@@ -791,43 +795,53 @@ export class CDPModule extends BaseModule {
+
+    const [collateralPoolBalances, totalSupply, cdpParams, tokenPriceAll, debtInfosAll, assetParamsAll, rateStrategies] = await Promise.all([collateralPoolBalancePromise, totalSupplyPromise, cdpParamsPromise, tokenPriceAllPromise, debtInfosPromise, assetParamsPromise, rateStrategyPromise]);
     const interestFee = bnOrZero(cdpParams.params?.interestFee);
     if (!interestFee) throw new Error("unable to retrieve interest fee");
     const tokenPrices = tokenPriceAll.tokenPrices;
     if (!tokenPrices) throw new Error("unable to retrieve token prices");
 
-    const cdpTokensBalancePromises: Promise<BigNumber>[] = cdpTokens.map(token => (
-      this.getCdpToActualRatio(token.denom, interestFee)
-        .then((ratio) => bnOrZero(token.amount).div(ratio))
-        .then(actualAmount => {
-          const denom = this.getUnderlyingDenom(token.denom);
-          const tokenPrice = tokenPrices.find((price) => price.denom === denom);
-          return this.getTokenUsdVal(denom, actualAmount, tokenPrice)
-        })
-        .catch((err) => {
-          console.error(err);
-          return BN_ZERO
-        })
-    ))
+    const cdpTokenBalances: Coin[] = (collateralPoolBalances?.balances ?? []).filter(balance => tokenClient.isCdpToken(balance.denom))
+
+    const cdpTokensBalancePromises = cdpTokenBalances.map(token => {
+      const underlyingDenom = this.getUnderlyingDenom(token.denom);
+      const tokenPrice = tokenPrices.find((price) => price.denom === underlyingDenom);
+      const supply = totalSupply.supply.find((supply) => supply.denom === token.denom)?.amount;
+      const balance = token.amount;
+      const debtInfo = debtInfosAll.debtInfosAll.find((debtInfo) => debtInfo.denom === underlyingDenom);
+
+      const assetParam = assetParamsAll.assetParamsAll.find((assetParam) => assetParam.denom === underlyingDenom);
+      const rateStrategy = rateStrategies.rateStrategyParamsAll.find((rateStrategy) => rateStrategy.name === assetParam?.rateStrategyName);
+
+      if (!debtInfo || !supply || !tokenPrice || !rateStrategy) throw new Error("unable to retrieve token info");
+
+      const apy = CDPModule.calculateInterestAPY(debtInfo, rateStrategy);
+      const newInterestRate = CDPModule.calculateInterestForTimePeriod(apy, debtInfo.lastUpdatedTime ?? new Date(0), new Date());
+      const principal = bnOrZero(debtInfo.totalPrincipal);
+      const accumInterest = bnOrZero(debtInfo.totalAccumulatedInterest);
+      const newInterest = principal.times(newInterestRate).plus(accumInterest.times(BN_ONE.plus(newInterestRate)));
+      const interest = newInterest.times(BN_10000.minus(interestFee)).dividedToIntegerBy(BN_10000);
+      const owedAmount = principal.plus(interest);
+      const totalOwed = bnOrZero(owedAmount).plus(bnOrZero(balance));
+      const ratio = bnOrZero(supply).div(bnOrZero(totalOwed));
+      const actualAmount = bnOrZero(balance).div(ratio);
+      return this.getTokenUsdVal(underlyingDenom, actualAmount, tokenPrice);
```

Comment:
Not sure if this is too much repeated code, given that similar logic is implemented in the other methods for single token. The alternative would be to reuse those methods but take in optional params given that they are calculated here, so for e.g. 

` public async getTotalTokenDebt(denom: string, debtInfo?: DebtInfo, interestFee?: BigNumber, newInterestRate?: BigNumber)`

Reply:
unsure which is "better design"

Reply:
Think the alternative (i.e. reuse the methods but take in optional params) is better.